contracts/Tokenomics.sol

```diff
@@ -888,18 +907,31 @@ contract Tokenomics is TokenomicsConstants {
 
         // Get the number of services
         uint256 numServices = serviceIds.length;
-        // Loop over service Ids to calculate their partial contributions
-        for (uint256 i = 0; i < numServices; ++i) {
-            // Check if the service owner or donator stakes enough OLAS for its components / agents to get a top-up
-            // If both component and agent owner top-up fractions are zero, there is no need to call external contract
-            // functions to check each service owner veOLAS balance
-            bool topUpEligible;
-            if (incentiveFlags[2] || incentiveFlags[3]) {
-                address serviceOwner = IToken(serviceRegistry).ownerOf(serviceIds[i]);
-                topUpEligible = (IVotingEscrow(ve).getVotes(serviceOwner) >= veOLASThreshold  ||
-                    IVotingEscrow(ve).getVotes(donator) >= veOLASThreshold) ? true : false;
+
+        // Check if the donator stakes enough OLAS for its components / agents to get a top-up
+        // If both component and agent owner top-up fractions are zero, there is no need to call external contract
+        // functions to check each service owner veOLAS balance
+        uint256 vPower;
+        if (incentiveFlags[2] || incentiveFlags[3]) {
+            vPower = IVotingEscrow(ve).getVotes(donator);
+            bytes32 donatorPowerHash = keccak256(abi.encode(epochCounter, donator));
+            // Check the donator voting power
+            if (vPower < veOLASThreshold || mapEpochDonatorPowerHashes[donatorPowerHash]) {
+                // If voting power is below the threshold or has been already utilized during the on-going epoch,
+                // top-ups are not eligible
+                vPower = 0;
+            } else {
+                // Increase the total top-up voting power in this epoch
+                mapDonationPoints[epochCounter].totalDonationPower += uint96(vPower);
```

Comment:
veOLAS is added to the `totalDonationPower` (per epoch) once for each donator who has not donated yet.